crates/mod-kafka/src/lib.rs

```diff
@@ -94,6 +96,69 @@ impl LuaUserData for Producer {
             Ok((partition, offset))
         });
 
+        methods.add_async_method("send_batch", |lua, this, values: Vec<Value>| async move {
+            let mut tasks = FuturesOrdered::new();
+            let producer = this.get_producer()?;
+
+            for value in values {
+                let record: Record = lua.from_value(value)?;
+
+                let headers = if record.headers.is_empty() {
+                    None
+                } else {
+                    let mut headers = OwnedHeaders::new();
+                    for (key, v) in &record.headers {
+                        headers = headers.insert(Header {
+                            key,
+                            value: Some(v),
+                        });
+                    }
+                    Some(headers)
+                };
+
+                let producer = producer.clone();
+
+                tasks.push_back(tokio::spawn(async move {
+                    producer
+                        .send(
+                            FutureRecord {
+                                topic: &record.topic,
+                                partition: record.partition,
+                                payload: record.payload.as_ref(),
+                                key: record.key.as_ref(),
+                                headers,
+                                timestamp: None,
+                            },
+                            Timeout::After(record.timeout.unwrap_or(Duration::from_secs(60))),
+                        )
+                        .await
+                }));
+            }
+
+            let mut failed_indexes = vec![];
+            let mut index = 1;
+
+            while let Some(result) = tasks.next().await {
+                match result {
+                    Ok(Ok(_)) => {}
+                    Ok(Err((error, _msg))) => {
+                        tracing::error!("Error sending to kafka {:?}", error);
+                        failed_indexes.push(index);
+                    }
+                    Err(error) => {
+                        tracing::error!("Error sending to kafka {:?}", error);
+                        failed_indexes.push(index)
+                    }
+                }
+                index += 1;
+            }
+            if failed_indexes.is_empty() {
+                Ok(Value::Nil)
+            } else {
+                Ok(lua.to_value(&failed_indexes)?)
+            }
```

Comment:
What do you think about doing it this way?

```suggestion
            let mut failed_indexes = vec![];
            let mut errors = vec![];
            let mut index = 1;

            while let Some(result) = tasks.next().await {
                match result {
                    Ok(Ok(_)) => {}
                    Ok(Err((error, _msg))) => {
                        errors.push(format!("{error:#}"));
                        failed_indexes.push(index);
                    }
                    Err(error) => {
                        errors.push(format!("{error:#}"));
                        failed_indexes.push(index);
                    }
                }
                index += 1;
            }
            Ok((failed_indexes, errors))
```

in the lua code you might do something like this:

```lua
local failed_items, errors = producer:send_batch(items)
if #failed_items > 0 then
   -- some items failed
   for i, item_idx in ipairs(failed_items) do
      local error = errors[i]
      print(string.format("item idx %d failed: %s", item_idx, error))
   end
end
```

I like not unilaterally emitting a `tracing::error` for this, because my assumption is that if any one in the batch is likely to fail, all of the batch is likely to fail and you can quickly flood the logs in that case.  Allowing you to control whether you log it from lua feels better.

Reply:
thanks for the feedback! I wasn't sure where the error should be logged, and you're right, having it in lua gives the user more control over it